### PR TITLE
Optimize zoning intersections

### PR DESCRIPTION
**Pull request by submarcos**

### CODE REVIEW

**Comment by submarcos** [comment]
Improve performances.

On a 260km tour : 
58s -> 21s trek detail page time generation
56s -> 6,3s SQL query execution
6Go -> 80Mo RAM consumed and kept

File: geotrek/common/utils/__init__.py [hunk 1]
```diff
@@ -101,16 +105,17 @@ def intersecting(qs, obj, distance=None, ordering=True, field='geom'):
...
-        qs = qs.filter(**{'{}__intersects'.format(field): obj.geom})
+        qs = qs.filter(Exists(obj._meta.model.objects.filter(geom__intersects=OuterRef('geom'), pk=obj.pk)))
         if obj.geom.geom_type == 'LineString' and ordering:
-            # FIXME: move transform from DRF viewset to DRF itself and remove transform here
-            ewkt = obj.geom.transform(settings.SRID, clone=True).ewkt
-            qs = qs.extra(select={'ordering': 'ST_LineLocatePoint(ST_GeomFromEWKT(\'{ewkt}\'), ST_StartPoint((ST_Dump(ST_Intersection(ST_GeomFromEWKT(\'{ewkt}\'), geom))).geom))'.format(ewkt=ewkt)})
-            qs = qs.extra(order_by=['ordering'])
+            qs = qs.order_by(LineLocatePoint(obj.geom,
+                                             StartPoint(DumpGeom(Intersection(obj.geom,
+                                                                              'geom')))))
```
Comment (submarcos) [inline geotrek/common/utils/__init__.py:112]:
```suggestion
                                                                              field)))))
```

**Comment by camillemonchicourt** [comment]
OK thanks, huge performance improvement.
Is it just on Geotrek-admin API, or also on Geotrek-admin web application ?
I guess the spatial indexes will improve performances everywhere ?

**Comment by submarcos** [comment]
> OK thanks, huge performance improvement. Is it just on Geotrek-admin API v2, or also on Geotrek-admin web application ? I guess the spatial indexes will improve performances everywhere ?

Because it concerns intersections on cities, districts and restricted areas, it has more impact on admin detail page. API v2 has only cities. For example trek detail api v2 endpoint : before 4,1s, after 3,64s

**Comment by camillemonchicourt** [comment]
OK, I see. Thanks.

**Review by Chatewgne — APPROVED** [review]

File: geotrek/zoning/mixins.py [hunk 2]
```diff
@@ -12,15 +13,15 @@ def zoning_property(self):
 
     @property
     def areas(self):
-        return uniquify(intersecting(RestrictedArea, self.zoning_property, distance=0))
+        return uniquify(intersecting(RestrictedArea, self.zoning_property, distance=0, defer=('geom',)))
```
Comment (numahell) [inline geotrek/zoning/mixins.py:16]: Question, qu'est-ce que ça fait le paramètre defer ?

**Review by numahell — APPROVED** [review]

File: geotrek/zoning/mixins.py (hunk 2 above)
Comment (submarcos) [reply geotrek/zoning/mixins.py:16]: https://github.com/GeotrekCE/Geotrek-admin/blob/optimize_zoning_intersections/geotrek/common/utils/__init__.py#L110

Comment (submarcos) [reply geotrek/zoning/mixins.py:16]: https://docs.djangoproject.com/fr/3.2/ref/models/querysets/#django.db.models.query.QuerySet.defer

**Pull request merged**